pkg/stats/collector.go

```diff
@@ -156,35 +162,78 @@ func (s *BufferedCollector) send(metrics []Metric) {
 	}
 }
 
+func (s *BufferedCollector) isClosed() bool {
+	return atomic.LoadInt32(&s.closed) == 1
+}
+
 func (s *BufferedCollector) CollectEvent(class, action string) {
+	if s.isClosed() {
+		return
+	}
+	s.wg.Add(1)
 	s.writes <- primaryKey{
 		class:  class,
 		action: action,
 	}
+	s.wg.Done()
 }
 
-func (s *BufferedCollector) Done() <-chan bool {
-	return s.done
+func (s *BufferedCollector) Close() {
+	// mark as closed and return in case it was closed
+	if !atomic.CompareAndSwapInt32(&s.closed, 0, 1) {
+		return
+	}
+	// wait for all events to be processed and close the channel
+	// done in the background because our main loop cloud be canceled
+	go func() {
+		s.wg.Wait()
+		close(s.writes)
+	}()
+	// wait for run loop to end
+	s.wgRun.Wait()
+	// drain channel and send cache for the last time
+	for w := range s.writes {
+		s.incr(w)
+	}
+	s.sendCache()
 }
 
 func (s *BufferedCollector) Run(ctx context.Context) {
-	go s.collectHeartbeat(ctx)
-	for {
-		select {
-		case w := <-s.writes: // collect events
-			s.incr(w)
-		case <-s.flushTicker.Tick(): // every N seconds, send the collected events
-			s.handleRuntimeStats()
-			metrics := makeMetrics(s.cache)
-			s.cache = make(keyIndex)
-			go s.send(metrics) // no need to block on this
-		case <-ctx.Done(): // we're done
-			metrics := makeMetrics(s.cache)
-			s.send(metrics)
-			s.done <- true
-			return
+	s.wgRun.Add(1)
+	go func() {
+		defer func() {
+			s.wgRun.Done()
+			// we trigger close after we mark that run ended
+			s.Close()
+		}()
+		for {
+			select {
+			case w, ok := <-s.writes: // collect events
+				if !ok {
+					return
+				}
+				s.incr(w)
+			case <-s.flushTicker.Tick(): // every N seconds, send the collected events
+				s.handleRuntimeStats()
+				s.sendCache()
+			case <-time.After(heartbeatInterval):
+				s.incr(primaryKey{class: "global", action: "heartbeat"})
+			case <-ctx.Done(): // we're done
+				return
+			}
 		}
+	}()
+}
+
+func (s *BufferedCollector) sendCache() {
+	if len(s.cache) == 0 {
+		return
+	}
+	metrics := makeMetrics(s.cache)
+	for k := range s.cache {
+		delete(s.cache, k)
```

Comment:
This is dangerous if someone added to the cache between the `metrics` were created and the cache was deleted